File: js/day2.js
let fs = require("fs");
const { parse } = require("path");

const input = fs.readFileSync("day2.txt", { encoding: "utf8" });
const lines = input.split("\r\n");
const games = lines.map((line) => {
  let [_, setsStr] = line.split(": ");
  return setsStr.split("; ").map((set) =>
    set.split(", ").map((cube) => {
      let [num, color] = cube.split(" ");
      return { color, num: parseInt(num) };
    })
  );
});

let silver = games.reduce(
  (acc, game, index) =>
    acc +
    (game.every((set) =>
      set.every((cube) => {
        switch (cube.color) {
          case "red":
            return cube.num <= 12;
          case "green":
            return cube.num <= 13;
          case "blue":
            return cube.num <= 14;
          default:
            console.warn("parsing error");
            return false;
        }
      })
    )
      ? index + 1
      : 0),
  0
);
console.log("silver", silver); // 2416

let gold = games.reduce((acc, game) => {
  let red = 0;
  let green = 0;
  let blue = 0;
  game.forEach((set) => {
    return set.forEach((cube) => {
      switch (cube.color) {
        case "red":
          red = Math.max(red, cube.num);
          break;
        case "green":
          green = Math.max(green, cube.num);
          break;
        case "blue":
          blue = Math.max(blue, cube.num);
          break;
        default:
          console.warn("parsing error");
      }
    });
  });
  return acc + red * green * blue;
}, 0);
console.log("gold", gold); // 63307

File: rb/day2.rb
require 'csv'

input = File.read('day2.txt')
lines = input.split("\n")
games = lines.map do |line|
  _, sets_str = line.split(': ')
  sets_str.split('; ').map do |set|
    set.split(', ').map do |cube|
      num, color = cube.split(' ')
      { color: color, num: num.to_i }
    end
  end
end

silver = games.each_with_index.reduce(0) do |acc, (game, index)|
  acc + (game.all? do |set|
    set.all? do |cube|
      case cube[:color]
      when 'red'
        cube[:num] <= 12
      when 'green'
        cube[:num] <= 13
      when 'blue'
        cube[:num] <= 14
      else
        puts 'parsing error'
        false
      end
    end
  end ? index + 1 : 0)
end

puts "silver #{silver}"

gold = games.reduce(0) do |acc, game|
  red = 0
  green = 0
  blue = 0
  game.each do |set|
    set.each do |cube|
      case cube[:color]
      when 'red'
        red = [red, cube[:num]].max
      when 'green'
        green = [green, cube[:num]].max
      when 'blue'
        blue = [blue, cube[:num]].max
      else
        puts 'parsing error'
      end
    end
  end
  acc + red * green * blue
end

puts "gold #{gold}"
